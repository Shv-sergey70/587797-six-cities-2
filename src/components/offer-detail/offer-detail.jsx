import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {MAX_RATING_VALUE} from "../../const/rating";
import Operation from '../../operation';
import ReviewItem from "../review-item/review-item";
import offerPropTypes from '../../prop-types/offer';
import Card from "../card/card";
import Selectors from '../../selector';
import Map from "../map/map";
import PageHeader from "../page-header/page-header";
import ReviewForm from "../review-form/review-form";
import {withReviewForm} from "../../hocs/with-review-form/with-review-form";
import {getRatingPercent} from "../../utils";
import commentPropTypes from '../../prop-types/comment';

const ReviewFormWrapped = withReviewForm(ReviewForm);

class OfferDetail extends React.PureComponent {
  constructor(props) {
    super(props);

    this._currentOfferId = Number(props.match.params.offerId);
    this._currentOffer = null;

    this._loadDataForOffer = this._loadDataForOffer.bind(this);
    this._getImagesForRender = this._getImagesForRender.bind(this);
    this._handleToggleFavoriteButtonClick = this._handleToggleFavoriteButtonClick.bind(this);
  }

  componentDidMount() {
    this._loadDataForOffer();
  }

  componentDidUpdate() {
    this._loadDataForOffer();
  }

  _loadDataForOffer() {
    const {
      loadOffersForDetailPage,
      currentOfferDetail,
      match: {
        params: {
          offerId
        }
      }
    } = this.props;

    this._currentOfferId = Number(offerId);
    this._currentOffer = currentOfferDetail;

    if (!this._currentOffer || this._currentOfferId !== currentOfferDetail.id) {
      loadOffersForDetailPage(this._currentOfferId);
    }
  }

  _getImagesForRender() {
    const {
      images
    } = this.props.currentOfferDetail;

    return images.length <= 6 ? images : images.slice(0, 6);
  }

  _handleToggleFavoriteButtonClick() {
    const {
      currentOfferDetail: {
        id,
        isFavorite
      },
      toggleFavoriteHotel
    } = this.props;

    toggleFavoriteHotel(id, !isFavorite);
  }

  render() {
    const {
      currentOfferDetail,
      currentComments,
      nearestOffers,
      isUserAuthorized
    } = this.props;

    if (currentOfferDetail === null) {
      return null;
    }

    const {
      title,
      bedrooms,
      maxAdults,
      goods,
      isPremium,
      isFavorite,
      type,
      price,
      rating,
      host,
      description
    } = currentOfferDetail;

    return <div className="page">
      <PageHeader/>
      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {this._getImagesForRender().map((imageSrc, i) => <div key={`image-${i}`} className="property__image-wrapper">
                <img className="property__image" src={imageSrc} alt="Photo studio"/>
              </div>)}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {isPremium
                ? <div className="property__mark">
                  <span>Premium</span>
                </div>
                : ``}
              <div className="property__name-wrapper">
                <h1 className="property__name">{title}</h1>
                <button className={`property__bookmark-button button ${isFavorite ? `place-card__bookmark-button--active` : ``}`} type="button"
                  onClick={this._handleToggleFavoriteButtonClick}>
                  <svg className="place-card__bookmark-icon" width="31" height="33">
                    <use xlinkHref="#icon-bookmark"/>
                  </svg>
                  <span className="visually-hidden">To bookmarks</span>
                </button>
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `${getRatingPercent(rating, MAX_RATING_VALUE)}%`}}> </span>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rating}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">{type}</li>
                <li className="property__feature property__feature--bedrooms">{bedrooms} Bedrooms</li>
                <li className="property__feature property__feature--adults">Max {maxAdults} adults</li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((item, i) => <li key={`goods-${i}`} className="property__inside-item">{item}</li>)}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                    <img className="property__avatar user__avatar" src={host.avatarUrl} width="74" height="74" alt="Host avatar"/>
                  </div>
                  <span className="property__user-name">{host.name}</span>
                  {host.isPro ? <span className="property__user-status">Pro</span> : ``}
                </div>
                <div className="property__description">
                  <p className="property__text">{description}</p>
                </div>
              </div>
              <section className="property__reviews reviews">
                <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{currentComments.length}</span></h2>
                <ul className="reviews__list">
                  {currentComments.map((comment) => <ReviewItem
                    key={`comment-${comment.id}`}
                    comment={comment}
                  />)}
                </ul>
                {isUserAuthorized ? <ReviewFormWrapped offerId={this._currentOfferId}/> : ``}
              </section>
            </div>
          </div>
          <section className="property__map map">
            <Map
              offersLocations={nearestOffers.map((offer) => offer.location)}
            />
          </section>
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <div className="near-places__list places__list">
              {nearestOffers.map((offer, i) => <Card
                key={`${offer.id}-${i}`}
                offer={offer}
                changeActiveOffer={() => {}}
              />)}
            </div>
          </section>
        </div>
      </main>
    </div>;
  }
}

OfferDetail.propTypes = {
  currentOfferDetail: PropTypes.object,
  match: PropTypes.shape({
    params: PropTypes.shape({
      offerId: PropTypes.string.isRequired
    })
  }),
  loadOffersForDetailPage: PropTypes.func.isRequired,
  currentComments: PropTypes.arrayOf(commentPropTypes),
  nearestOffers: PropTypes.arrayOf(offerPropTypes),
  toggleFavoriteHotel: PropTypes.func.isRequired,
  isUserAuthorized: PropTypes.bool.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  isUserAuthorized: Selectors.isUserAuthorized(state),
  currentOfferDetail: state.currentOfferDetail,
  currentComments: Selectors.getSortedCurrentComments(state),
  nearestOffers: Selectors.getNearestOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadOffersForDetailPage: (currentOffer) => dispatch(Operation.loadOffersForDetailPage(currentOffer)),
  toggleFavoriteHotel: (offerId, isSetFavorite) => dispatch(Operation.toggleFavoriteHotel(offerId, isSetFavorite))
});

export {OfferDetail};
export default connect(mapStateToProps, mapDispatchToProps)(OfferDetail);
