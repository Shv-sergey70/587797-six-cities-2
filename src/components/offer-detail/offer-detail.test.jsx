import React from 'react';
import renderer from 'react-test-renderer';
import {OfferDetail} from "./offer-detail";

jest.mock(`../page-header/page-header.jsx`, () => `PageHeader`);
jest.mock(`../card/card.jsx`, () => `Card`);
jest.mock(`../review-item/review-item.jsx`, () => `ReviewItem`);
jest.mock(`../map/map.jsx`, () => `Map`);
jest.mock(`../review-form/review-form.jsx`, () => `ReviewForm`);

jest.mock(`../../notifier`, () => ({
  error: jest.fn(),
}));

const nearestOffersData = [
  {
    id: 1,
    title: `Beautiful & luxurious apartment at great location`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: true,
    isFavorite: true,
    type: `apartment`,
    price: 80,
    rating: 4.4,
    bedrooms: 2,
    maxAdults: 2,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`, `/img/studio-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    description: `A quiet cozy and picturesque that hides behind a a river`
  },
  {
    id: 2,
    title: `Beautiful & luxurious apartment at great location2`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: false,
    isFavorite: true,
    type: `apartment`,
    price: 180,
    rating: 2.4,
    bedrooms: 5,
    maxAdults: 3,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`, `/img/studio-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`],
    description: `A quiet cozy and picturesque that hides behind a a river1`
  },
  {
    id: 3,
    title: `Beautiful & luxurious apartment at great location33`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: true,
    isFavorite: false,
    type: `room`,
    price: 820,
    rating: 4.9,
    bedrooms: 5,
    maxAdults: 4,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`],
    description: `A quiet cozy and picturesque that hides behind a a river22`
  }
];

const currentComments = [
  {
    id: 1,
    user: {
      id: 1,
      isPro: true,
      name: `Angelina`,
      avatarUrl: `/img/avatar-angelina.jpg`
    },
    rating: 3,
    comment: `Comment`,
    date: `2019-12-12`
  },
  {
    id: 2,
    user: {
      id: 2,
      isPro: true,
      name: `Angelina`,
      avatarUrl: `/img/avatar-angelina.jpg`
    },
    rating: 4,
    comment: `Comment2`,
    date: `2019-12-15`
  }
];

it(`Offer detail correctly renders`, () => {
  const tree = renderer.create(
      <OfferDetail
        currentOfferDetail = {
          {
            id: 1,
            title: `Beautiful & luxurious apartment at great location`,
            images: [`/img/room.jpg`, `/img/apartment-01.jpg`, `/img/apartment-01.jpg`, `/img/studio-01.jpg`],
            bedrooms: 2,
            maxAdults: 2,
            goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
            isPremium: true,
            isFavorite: true,
            type: `Apartment`,
            price: 80,
            rating: 4.5,
            host: {
              isPro: true,
              name: `Angelina`,
              avatarUrl: `/img/avatar-angelina.jpg`
            },
            description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`
          }
        }
        match={{
          params: {
            offerId: `1`
          }
        }}
        loadOffersForDetailPage={jest.fn()}
        currentComments={currentComments}
        nearestOffers={nearestOffersData}
        toggleFavoriteHotel={jest.fn()}
        isUserAuthorized={true}
      />).toJSON();

  expect(tree).toMatchSnapshot();
});
