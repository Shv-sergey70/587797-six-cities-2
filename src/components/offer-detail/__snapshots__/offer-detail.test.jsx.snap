// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offer detail correctly renders 1`] = `
<div
  className="page"
>
  <PageHeader />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/room.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-01.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-01.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/studio-01.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span>
              Premium
            </span>
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious apartment at great location
            </h1>
            <button
              className="property__bookmark-button button place-card__bookmark-button--active"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                 
              </span>
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              4.5
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              Apartment
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              2
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              2
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              80
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="/img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                2
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <ReviewItem
                comment={
                  Object {
                    "comment": "Comment",
                    "date": "2019-12-12",
                    "id": 1,
                    "rating": 3,
                    "user": Object {
                      "avatarUrl": "/img/avatar-angelina.jpg",
                      "id": 1,
                      "isPro": true,
                      "name": "Angelina",
                    },
                  }
                }
              />
              <ReviewItem
                comment={
                  Object {
                    "comment": "Comment2",
                    "date": "2019-12-15",
                    "id": 2,
                    "rating": 4,
                    "user": Object {
                      "avatarUrl": "/img/avatar-angelina.jpg",
                      "id": 2,
                      "isPro": true,
                      "name": "Angelina",
                    },
                  }
                }
              />
            </ul>
            <ReviewForm
              offerId={1}
              onCommentChange={[Function]}
              onRatingClick={[Function]}
              rating={null}
              resetForm={[Function]}
              review=""
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <Map
          offersLocations={
            Array [
              undefined,
              undefined,
              undefined,
            ]
          }
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        >
          <Card
            changeActiveOffer={[Function]}
            offer={
              Object {
                "bedrooms": 2,
                "description": "A quiet cozy and picturesque that hides behind a a river",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                  "Coffee machine",
                  "Dishwasher",
                ],
                "id": 1,
                "images": Array [
                  "/img/room.jpg",
                  "/img/apartment-01.jpg",
                  "/img/studio-01.jpg",
                ],
                "isFavorite": true,
                "isPremium": true,
                "maxAdults": 2,
                "previewImage": "/img/apartment-01.jpg",
                "price": 80,
                "rating": 4.4,
                "title": "Beautiful & luxurious apartment at great location",
                "type": "apartment",
              }
            }
          />
          <Card
            changeActiveOffer={[Function]}
            offer={
              Object {
                "bedrooms": 5,
                "description": "A quiet cozy and picturesque that hides behind a a river1",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                  "Coffee machine",
                ],
                "id": 2,
                "images": Array [
                  "/img/room.jpg",
                  "/img/apartment-01.jpg",
                  "/img/studio-01.jpg",
                ],
                "isFavorite": true,
                "isPremium": false,
                "maxAdults": 3,
                "previewImage": "/img/apartment-01.jpg",
                "price": 180,
                "rating": 2.4,
                "title": "Beautiful & luxurious apartment at great location2",
                "type": "apartment",
              }
            }
          />
          <Card
            changeActiveOffer={[Function]}
            offer={
              Object {
                "bedrooms": 5,
                "description": "A quiet cozy and picturesque that hides behind a a river22",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                ],
                "id": 3,
                "images": Array [
                  "/img/room.jpg",
                  "/img/apartment-01.jpg",
                ],
                "isFavorite": false,
                "isPremium": true,
                "maxAdults": 4,
                "previewImage": "/img/apartment-01.jpg",
                "price": 820,
                "rating": 4.9,
                "title": "Beautiful & luxurious apartment at great location33",
                "type": "room",
              }
            }
          />
        </div>
      </section>
    </div>
  </main>
</div>
`;
